agent/proxycfg/manager.go

```diff
@@ -0,0 +1,296 @@
+			// registered and once we get past here, we don't have enough context to
+			// know that so we'd need to set it here if not during registration of the
+			// proxy service. Sidecar Service and managed proxies in the interim can
+			// do that, but we should validate more generally that that is always
+			// true.
+			err := m.ensureProxyServiceLocked(svc, m.cfg.State.ServiceToken(svcID))
+			if err != nil {
+				m.cfg.Logger.Printf("[ERR] failed to watch proxy service %s: %s", svc.ID,
+					err)
+			}
+		}
+
+		// Now see if any proxies were removed
+		for proxyID := range m.proxies {
+			if _, ok := services[proxyID]; !ok {
+				// Remove them
+				m.removeProxyServiceLocked(proxyID)
+			}
+		}
+
+		m.mu.Unlock()
+
+		// Wait for a state change
+		_, ok := <-stateCh
+		if !ok {
+			// Stopped
+			return nil
+		}
+	}
+}
+
+// ensureProxyServiceLocked adss or changes the proxy to our state.
+func (m *Manager) ensureProxyServiceLocked(ns *structs.NodeService, token string) error {
+	state, ok := m.proxies[ns.ID]
+
+	if ok {
+		if !state.Changed(ns, token) {
+			// No change
+			return nil
+		}
+
+		// We are updating the proxy, close it's old state
+		state.Close()
+	}
+
+	var err error
+	state, err = newState(ns, token)
+	if err != nil {
+		return err
+	}
+
+	// Set the necessary dependencies
+	state.logger = m.cfg.Logger
+	state.cache = m.cfg.Cache
+	state.source = m.cfg.Source
+
+	ch, err := state.Watch()
+	if err != nil {
+		return err
+	}
+	m.proxies[ns.ID] = state
+
+	// Start a goroutine that will wait for changes and broadcast them to watchers.
+	go func(ch <-chan ConfigSnapshot) {
+		// Run until ch is closed
+		for snap := range ch {
+			m.notify(&snap)
+		}
+	}(ch)
+
+	return nil
+}
+
+// removeProxyService is called when a service deregisteres and frees all
+// resources for that service.
+func (m *Manager) removeProxyServiceLocked(proxyID string) {
+	state, ok := m.proxies[proxyID]
+	if !ok {
+		return
+	}
+
+	// Closing state will let the goroutine we started in Ensure finish since
+	// watch chan is closed.
+	state.Close()
+	delete(m.proxies, proxyID)
+
+	// We intentionally leave potential watchers hanging here - there is no new
+	// config for them and closing their channels might be indistinguishable from
+	// an error that they should retry. We rely for them to eventually give up
+	// (because they are in fact not running any more) and so the watches be
+	// cleaned up naturally.
+}
+
+func (m *Manager) notify(snap *ConfigSnapshot) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	watchers, ok := m.watchers[snap.ProxyID]
+	if !ok {
+		return
+	}
+
+	for _, ch := range watchers {
+		// Attempt delivery but don't let slow consumers block us forever. They
+		// might miss updates but it's better than breaking everything.
+		//
+		// TODO(banks): should we close their chan here to force them to eventually
+		// notice they are too slow? Not sure if it really helps.
+		select {
+		case ch <- snap:
```

Comment:
If the intent is to try and always deliver the _latest_ copy of the snapshot to the watcher on the 1-buffered channels, you could do something like:

```
// sends snap into ch and replaces the item in the channel if one is already present.
// assumes channel is buffered with a capacity of 1 and there is only one sender.
func bumpsend(ch chan *ConfigSnapshot, snap *ConfigSnapshot) {
	// optimistically assume buffer is empty
	select {
	case ch <- snap:
		return
	default:
	}

	// drain the stale snapshot
	select {
	case <-ch:
	default:
	}

	// replace the head
	select {
	case ch <- snap:
		return
	default:
		panic("there are two senders")
	}
}
```